### PR TITLE
Add edge-coloured digraph automorphism finding method

### PR DESCRIPTION
**Pull request by flsmith**
Resolves https://github.com/gap-packages/Digraphs/issues/183.

### CODE REVIEW

File: gap/isomorph.gi [hunk 1]
```diff
@@ -561,6 +612,140 @@ function(n, partition, method)
...
+    for i in adj_colours do
+      if not IsPosInt(i) then
+        ErrorNoReturn("the 2nd argument should be a list of lists of",
+                      "positive integers");
+      fi;
+      AddSet(colours, i);
+    od;
+  od;
+  m := Length(colours);
+  if ForAny([1 .. m], i -> i <> colours[i]) then
```
Comment (ChrisJefferson) [inline gap/isomorph.gi]: It won't make much difference, but you could write `if colours = [1..m] ...` here.

**Comment by markuspf** [comment]
It looks like the colouring consistency check takes a really long time: I made an example with 3500 vertices and it doesn't get to computing the autmorphism group at all.

**Comment by wilfwilson** [comment]
Thanks Markus, really useful observation. Hopefully we’ll be able to sort that out.

**Comment by flsmith** [comment]
Fix will come tonight :)

**Comment by james-d-mitchell** [comment]
Any chance you could fix the CI @flsmith and rebase onto the current `master` branch??

**Comment by flsmith** [comment]
I realised recently that this reduction is invalid for multidigraphs (even with unique edge colours) and haven't found a way to fix it efficiently yet.

**Comment by ChrisJefferson** [comment]
When I've done something similar in the past, I first made a function from edge-labelled multidigraphs to edge-labelled digraphs, by making a pass over the graph and making new a new label whenever I find a new multiset of edges. I think a function that did that might have other uses, so it's probably not a silly thing to write anyway.

It's probably not possible to do this (easily) in a single pass -- you possibly could do something crazy, where you only add a new layer when you realise you need one as the number of edge labels grows, but I certainly wouldn't do that as a first pass

**Comment by james-d-mitchell** [comment]
Cough cough @flsmith

**Comment by james-d-mitchell** [comment]
For the record I think this should be rebased, and then hopefully the CI will pass.

**Comment by james-d-mitchell** [comment]
I'm happy with this, and think it should be merged sooner rather than later.

File: .travis.yml [hunk 2]
```diff
@@ -82,8 +82,8 @@ matrix:
       - ABI=64
       - GAP=master
       - PACKAGES=required
-      - THRESHOLD=98
-
+      - THRESHOLD=94
```
Comment (wilfwilson) [inline .travis.yml]: This is... unfortunate. Is there a deep reason why you can't get better code coverage in `isomorph.gi` via `isomorph.tst`, or is it just that you've not got round to it?

If you're not going to get the coverage up, I'd prefer you to put this back up to `98` anyway, and instead put `isomorph.tst` into `.covignore`.

File: doc/isomorph.xml [hunk 3]
```diff
@@ -41,7 +41,7 @@
 
 <#GAPDoc Label="BlissAutomorphismGroup">
 <ManSection>
-  <Attr Name="BlissAutomorphismGroup" Arg="digraph[, colours]"/>
+  <Attr Name="BlissAutomorphismGroup" Arg="digraph[, vertex_colours, edge_colours]"/>
```
Comment (wilfwilson) [inline doc/isomorph.xml]: Does this have to be used with either both optional arguments or none, or is just vertex colours okay? In which case:
```suggestion
  <Attr Name="BlissAutomorphismGroup" Arg="digraph[, vertex_colours[, edge_colours]]"/>
```

File: doc/isomorph.xml [hunk 4]
```diff
@@ -268,23 +206,93 @@ true
...
+      <Item>
+        <A>edge_colours</A> must be fail or a list of <C>n</C> lists of integers
+        of the same shape as <C>OutNeighbours(digraph)</C>, where
+        <A>edge_colours</A><C>[i][j]</C> is the colour of the edge
+        <C>OutNeighbours(digraph)[i][j]</C>, using the colours <C>[1 .. k]</C>
+        for some <C>k &lt;= n</C>;
+      </Item>
+    </List>
+
+    Gving <A>vert_colours</A> [<A>edge_colours</A>] as <C>fail</C> is equivalent
```
Comment (wilfwilson) [inline doc/isomorph.xml]:
```suggestion
    Giving <A>vert_colours</A> [<A>edge_colours</A>] as <C>fail</C> is equivalent
```

File: doc/isomorph.xml [hunk 5]
```diff
@@ -54,6 +54,9 @@
     the value of either <C>BlissAutomorphismGroup</C> or  <Ref
       Attr="NautyAutomorphismGroup"/>. These groups are, of course, equal but
     their generating sets may differ. <P/>
+    The attribute <Ref Attr="AutomorphismGroup" Label="for a digraph,
+      homogeneous list, and list" /> returns the value of
+    <C>BlissAutomorphismGroup</C> as it is not implemented for &nauty;
```
Comment (wilfwilson) [inline doc/isomorph.xml]: Perhaps say that that is where you can find the requirements of the optional arguments to this function?

Also, the three-argument version of `BlissAutomorphismGroup` is not an attribute, so I think it should be documented more like:
```suggestion
  <Attr Name="BlissAutomorphismGroup" Arg="digraph"/>
  <Oper Name="BlissAutomorphismGroup" Arg="digraph, vertex_colours, edge_colours"/>
```
I presume - or something like that (perhaps with a two-argument version, if applicable, as mentioned above). Perhaps they will need labels.

File: doc/isomorph.xml [hunk 6]
```diff
@@ -54,6 +54,9 @@
     the value of either <C>BlissAutomorphismGroup</C> or  <Ref
       Attr="NautyAutomorphismGroup"/>. These groups are, of course, equal but
     their generating sets may differ. <P/>
+    The attribute <Ref Attr="AutomorphismGroup" Label="for a digraph,
```
Comment (wilfwilson) [inline doc/isomorph.xml]: You can't have line-breaks in the label bit of a `Ref` tag - GAP doc gives warnings.

File: doc/isomorph.xml [hunk 7]
```diff
@@ -268,23 +206,93 @@ true
...
+    for more information about isomorphisms of coloured digraphs. <P/>
+
+    If <A>digraph</A> is not a multidigraph then the automorphism group is
+    returned as a group of permutations on the set of vertices of
+    <A>digraph</A>. <P/>
+
+    If <A>digraph</A> is a multidigraph then the automorphism group is returned
+    as the direct product of a group of permutations on the set of vertices of
+    <A>digraph</A> with a group of permutations on the set of edges of
+    <A>digraph</A>. These groups can be accessed using <Ref Attr="Projection"/>
```
Comment (wilfwilson) [inline doc/isomorph.xml]: All references to `Projection` should probably be
```
<Ref Oper="Projection" Label="for a domain and a positive integer" BookName="ref"/>
```

File: doc/isomorph.xml [hunk 8]
```diff
@@ -268,23 +206,93 @@ true
...
+    This operation computes the automorphism group of a vertex- and/or
+    edge-coloured digraph.
+    A coloured digraph can be specified by its underlying digraph
+    <A>digraph</A> and colourings <A>vert_colours</A>, <A>edge_colours</A>.
+    Let <C>n</C> be the number of vertices of <A>digraph</A>. The colourings
+    must have the following forms:
+
+    <List>
+      <Item>
+        <A>vert_colours</A> must be <C>fail</C> or a list of <C>n</C> integers,
```
Comment (wilfwilson) [inline doc/isomorph.xml]:
```suggestion
        <A>vert_colours</A> must be <K>fail</K> or a list of <C>n</C> integers,
```

File: doc/isomorph.xml [hunk 9]
```diff
@@ -268,23 +206,93 @@ true
...
+        where <A>vert_colours</A><C>[i]</C> is the colour of vertex <C>i</C>,
+        using the colours <C>[1 .. m]</C> for some <C>m &lt;= n</C>;
+      </Item>
+      <Item>
+        <A>edge_colours</A> must be fail or a list of <C>n</C> lists of integers
```
Comment (wilfwilson) [inline doc/isomorph.xml]:
```suggestion
        <A>edge_colours</A> must be <K>fail</K> or a list of <C>n</C> lists of integers
```

File: .travis.yml (hunk 2 above)
Comment (james-d-mitchell) [reply .travis.yml]: I'm not sure what is going on here, but @flsmith agreed to try to remove this change yesterday, if I remember correctly.

Comment (flsmith) [reply .travis.yml]: Caused by my confusion about nauty and now fixed

File: doc/isomorph.xml (hunk 3 above)
Comment (flsmith) [reply doc/isomorph.xml]: Just vertex colour is okay, fixed

File: doc/isomorph.xml (hunk 5 above)
Comment (flsmith) [reply doc/isomorph.xml]: Agreed and fixed

**Pull request merged**